File: src/TabulaSonora.Web/wwwroot/js/asset-store.js
// Persistent storage for the two files the engine needs and the page cannot ship: the user's own
// SCCore.dll, and optionally their own presets.json.
//
// The DLL is 27 MB. Reading it and writing it stays entirely on this side -- the bytes never travel
// through Blazor's InputFile stream, which is the wrong shape for a file this size. .NET asks for
// them once, into a buffer it has already sized, through readInto below.

const DB_NAME = 'tabula-sonora';
const DB_VERSION = 1;
const STORE = 'assets';

let cached = new Map();

function open() {
    return new Promise((resolve, reject) => {
        const request = indexedDB.open(DB_NAME, DB_VERSION);
        request.onupgradeneeded = () => request.result.createObjectStore(STORE);
        request.onsuccess = () => resolve(request.result);
        request.onerror = () => reject(request.error);
    });
}

function transact(db, mode, work) {
    return new Promise((resolve, reject) => {
        const tx = db.transaction(STORE, mode);
        const request = work(tx.objectStore(STORE));
        tx.oncomplete = () => resolve(request ? request.result : undefined);
        tx.onerror = () => reject(tx.error);
        tx.onabort = () => reject(tx.error);
    });
}

async function sha256Hex(buffer) {
    const digest = await crypto.subtle.digest('SHA-256', buffer);
    return Array.from(new Uint8Array(digest))
        .map(b => b.toString(16).padStart(2, '0'))
        .join('');
}

// Ask the browser not to evict us. Without it a 27 MB record is "best effort" storage and can be
// cleared under disk pressure, which would silently send the user back to the file picker.
export async function requestPersistence() {
    if (!navigator.storage?.persist) {
        return false;
    }

    return (await navigator.storage.persisted()) || (await navigator.storage.persist());
}

// Metadata only. The bytes stay in IndexedDB until .NET actually asks for them, so a page load that
// finds nothing cached costs one small read rather than 27 MB of decoding.
export async function describe(key) {
    const db = await open();
    try {
        const record = await transact(db, 'readonly', store => store.get(key));
        if (!record) {
            return null;
        }

        return {
            key,
            name: record.name,
            size: record.bytes.byteLength,
            sha256: record.sha256,
            savedAt: record.savedAt,
        };
    } finally {
        db.close();
    }
}

// Reads the file sitting in a picker, hashes it, and stores it. Taking the element rather than a
// Blazor IBrowserFile is what keeps 27 MB off the interop path twice over: the bytes go from the
// File straight into IndexedDB, and .NET only ever sees them once, on the way into the engine.
export async function putFromInput(key, element) {
    const file = element?.files?.[0];
    if (!file) {
        return null;
    }

    const bytes = await file.arrayBuffer();
    const sha256 = await sha256Hex(bytes);
    const record = { name: file.name, bytes, sha256, savedAt: new Date().toISOString() };

    const db = await open();
    try {
        await transact(db, 'readwrite', store => store.put(record, key));
    } finally {
        db.close();
    }

    cached.set(key, bytes);
    return { key, name: record.name, size: bytes.byteLength, sha256, savedAt: record.savedAt };
}

export async function remove(key) {
    cached.delete(key);
    const db = await open();
    try {
        await transact(db, 'readwrite', store => store.delete(key));
    } finally {
        db.close();
    }
}

// Hands the stored bytes to .NET, which reads them into a buffer of its own.
//
// Returns the array itself, NOT DotNet.createJSStreamReference(it): when the .NET side declares the
// result as IJSStreamReference, Blazor calls that for us on whatever comes back. Wrapping it here as
// well hands it a plain object to wrap a second time, which fails with "Supplied value is not a
// typed array or blob".
export async function readStream(key) {
    let bytes = cached.get(key);
    if (!bytes) {
        const db = await open();
        try {
            const record = await transact(db, 'readonly', store => store.get(key));

            // Empty rather than null: Blazor dereferences whatever comes back on its way to a stream
            // reference, so returning null would throw inside the framework instead of here.
            if (!record) {
                return new Uint8Array(0);
            }

            bytes = record.bytes;
        } finally {
            db.close();
        }

        cached.set(key, bytes);
    }

    return new Uint8Array(bytes);
}

// The 27 MB stays referenced above so a retry does not re-read IndexedDB, but it is dead weight once
// .NET has its own copy. Called after the ROM is handed over.
export function releaseCache(key) {
    cached.delete(key);
}

export async function readText(key) {
    const db = await open();
    try {
        const record = await transact(db, 'readonly', store => store.get(key));
        return record ? new TextDecoder().decode(record.bytes) : null;
    } finally {
        db.close();
    }
}

File: src/TabulaSonora.Web/Services/AssetStore.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace TabulaSonora.Web.Services;

/// <summary>What is held in browser storage under one key.</summary>
/// <param name="Key">Storage key.</param>
/// <param name="Name">File name as the user picked it.</param>
/// <param name="Size">Length in bytes.</param>
/// <param name="Sha256">SHA-256 as lower-case hex, computed when it was stored.</param>
/// <param name="SavedAt">When it was stored, ISO-8601.</param>
public sealed record StoredAsset(string Key, string Name, long Size, string Sha256, string SavedAt);

/// <summary>
/// The user's own files, kept in IndexedDB so they are supplied once rather than every visit.
/// </summary>
/// <remarks>
/// Nothing Roland's is served by this application; the DLL is the user's, it stays on their machine,
/// and it is never uploaded anywhere. This class is the whole of its lifetime in the browser.
/// </remarks>
public sealed class AssetStore(IJSRuntime js) : IAsyncDisposable
{
    /// <summary>Storage key for <c>SCCore.dll</c>.</summary>
    public const string RomKey = "sccore";

    /// <summary>Storage key for a user-supplied <c>presets.json</c>.</summary>
    public const string PresetsKey = "presets";

    /// <summary>Upper bound on a stream read out of storage; the DLL is 27 MB.</summary>
    private const long MaxAssetBytes = 64L * 1024 * 1024;

    private IJSObjectReference? _module;

    private async ValueTask<IJSObjectReference> ModuleAsync() =>
        _module ??= await js.InvokeAsync<IJSObjectReference>("import", "./js/asset-store.js");

    /// <summary>Asks the browser not to evict this origin's storage.</summary>
    /// <returns>Whether storage is now persistent.</returns>
    /// <remarks>
    /// Without the grant a 27 MB record is best-effort and can be cleared under disk pressure, which
    /// would silently send the user back to the file picker with no explanation.
    /// </remarks>
    public async ValueTask<bool> RequestPersistenceAsync() =>
        await (await ModuleAsync()).InvokeAsync<bool>("requestPersistence");

    /// <summary>Reads what is stored under a key, without reading its bytes.</summary>
    /// <param name="key">Storage key.</param>
    /// <returns>The record, or <see langword="null"/> if nothing is stored.</returns>
    public async ValueTask<StoredAsset?> DescribeAsync(string key) =>
        await (await ModuleAsync()).InvokeAsync<StoredAsset?>("describe", key);

    /// <summary>Stores the file currently selected in a file input.</summary>
    /// <param name="key">Storage key.</param>
    /// <param name="input">The <c>&lt;input type="file"&gt;</c> element.</param>
    /// <returns>The stored record, or <see langword="null"/> if nothing was selected.</returns>
    /// <remarks>
    /// Takes the element rather than a <c>IBrowserFile</c> on purpose: the bytes go from the picked
    /// file straight into IndexedDB without crossing into .NET at all, which for the DLL is 27 MB of
    /// interop that never has to happen.
    /// </remarks>
    public async ValueTask<StoredAsset?> PutFromInputAsync(string key, ElementReference input) =>
        await (await ModuleAsync()).InvokeAsync<StoredAsset?>("putFromInput", key, input);

    /// <summary>Deletes what is stored under a key.</summary>
    /// <param name="key">Storage key.</param>
    public async ValueTask RemoveAsync(string key) =>
        await (await ModuleAsync()).InvokeVoidAsync("remove", key);

    /// <summary>Reads a stored asset's bytes.</summary>
    /// <param name="key">Storage key.</param>
    /// <returns>The bytes, or <see langword="null"/> if nothing is stored.</returns>
    public async ValueTask<byte[]?> ReadAsync(string key)
    {
        var module = await ModuleAsync();
        var reference = await module.InvokeAsync<IJSStreamReference>("readStream", key);

        await using (reference)
        {
            // Nothing stored under that key comes back as an empty array rather than null, because
            // Blazor dereferences the result on its way to a stream reference.
            if (reference.Length == 0)
            {
                return null;
            }

            await using var stream = await reference.OpenReadStreamAsync(MaxAssetBytes);
            using var buffer = new MemoryStream((int)Math.Min(reference.Length, MaxAssetBytes));
            await stream.CopyToAsync(buffer);

            // The browser side keeps its own reference alive for retries; once the bytes are here
            // that is a second 27 MB nobody needs.
            await module.InvokeVoidAsync("releaseCache", key);
            return buffer.ToArray();
        }
    }

    /// <summary>Reads a stored asset as UTF-8 text.</summary>
    /// <param name="key">Storage key.</param>
    /// <returns>The text, or <see langword="null"/> if nothing is stored.</returns>
    public async ValueTask<string?> ReadTextAsync(string key) =>
        await (await ModuleAsync()).InvokeAsync<string?>("readText", key);

    /// <summary>Releases the JavaScript module.</summary>
    public async ValueTask DisposeAsync()
    {
        if (_module is not null)
        {
            await _module.DisposeAsync();
            _module = null;
        }
    }
}
